### PR TITLE
Update librustzcash

### PR DESCRIPTION
**Pull request by Duddino**
Updates librustzcash to version 0.11.0
I'm depending on [a modified version](https://github.com/Duddino/librustzcash/tree/pivx), this is because some methods are not made public by upstream and to strip Orchard since its license is incompatible with MIT

### CODE REVIEW

File: src/rust/src/rustzcash.rs [hunk 1]
```diff
@@ -514,24 +457,25 @@ pub extern "system" fn librustzcash_sapling_ka_agree(
...
     };
 
-    // Compute key agreement
-    let ka = sapling_ka_agree(&sk, &p);
-
-    // Produce result
-    let result = unsafe { &mut *result };
-    ka.write(&mut result[..]).expect("length is not 32 bytes");
+    let shared_secret = SaplingDomain::<TestNetwork>::ka_agree_enc(&esk, &p);
+    *&mut unsafe { *result } = shared_secret.0.to_bytes();
```
Comment (panleone) [inline src/rust/src/rustzcash.rs:478]: trivial bug (in this way result is not updated)
a possible fix is:
`unsafe{*result = shared_secret.0.to_bytes()};`

**Review by panleone — CHANGES_REQUESTED** [review]
Impressive work! Spent most of the day testing the PR  here the results:
- I managed to sync a new node from 0
- All type of transactions are successfully created and accepted by peers

So overall the code is working, I also started testing something in more details and I found a couple of bugs:
- The shared secret used to decrypt sapling notes is not generated properly (see the other comment for a possible fix)
- The ivk key is also not generated as it should be (but I still have to investigate on this bug)

**Comment by panleone** [comment]
Update: just figured out that the suggested fix solves both of the problems I reported

File: src/rust/src/rustzcash.rs (hunk 1 above)
Comment (Duddino) [reply src/rust/src/rustzcash.rs]: It's weird that it wouldn't work. I've applied your suggestion.

File: src/rust/src/rustzcash.rs [hunk 2]
```diff
@@ -282,39 +243,35 @@ pub extern "system" fn librustzcash_to_scalar(
...
     // size of the representation
-    let scalar = <Bls12 as JubjubEngine>::Fs::to_uniform(unsafe { &(&*input)[..] }).into_repr();
-
-    let result = unsafe { &mut *result };
+    let scalar = Fr::from_bytes_wide(&mut unsafe { *input });
 
-    scalar
-        .write_le(&mut result[..])
-        .expect("length is 32 bytes");
+    *&mut unsafe { *result } = scalar.to_bytes();
```
Comment (panleone) [inline src/rust/src/rustzcash.rs:248]: change this line to:
unsafe{*result = scalar.to_bytes()};
or result is not updated

File: src/rust/src/rustzcash.rs [hunk 3]
```diff
@@ -1191,9 +1176,10 @@ pub extern "system" fn librustzcash_zip32_xfvk_derive(
         Err(_) => return false,
     };
 
-    xfvk.write(&mut (unsafe { &mut *xfvk_i })[..])
+    let mut bytes = [0u8; 169];
+    xfvk.write(&mut bytes.as_mut_slice())
         .expect("should be able to serialize an ExtendedFullViewingKey");
-
+    *&mut unsafe { *xfvk_i } = bytes;
```
Comment (panleone) [inline src/rust/src/rustzcash.rs:1182]: change this line to:
unsafe{*xfvk_i = bytes};
or xfvk_i is not updated

**Review by panleone — COMMENTED** [review]
Spent another day testing this PR:
nice job on removing orchard from the lib (it works flawless)
everything is retrocompatible with the current main, unit and functional test pass with the only 2 changes: (see the 2 other comments)

Doesn't compile here on github because CI is in offline mode and cannot fetch the lib from your github, but it should be easily fixable

**Review by panleone — APPROVED** [review]
tACK 05d7993caf9ae1141256e9598c7ba4f918adccae let's get this merged

**Review by Fuzzbawls — COMMENTED** [review]
Looks like the `Cargo.lock` file was edited manually or by a diff patch instead of running `cargo update`? This has resulted in including packages/crates that we won't be using...and are specifically avoiding. Even unused packages get downloaded/compiled, and in the case of depends builds, vendored.

This is what `cargo update` shows for me if it is of any use:
Image 059

Also of note: This renders the lib's internal unit tests inoperable, which are currently passing on `master`

File: Cargo.lock [hunk 4]
```diff
@@ -1,679 +1,1268 @@
...
+checksum = "f0f9ef7462f7c099f518d754361858f86d8a07af53ba9af0fe635bbccb151a63"
+dependencies = [
+ "ff",
+ "memuse",
+ "rand_core",
+ "subtle",
+]
+
+[[package]]
+name = "halo2_gadgets"
```
[the remark this answers is not in the document]
Comment (Duddino) [reply Cargo.lock]: Seems like Cargo.lock was overridden by a previous version, should be fixed now

**Review by panleone — APPROVED** [review]
re-tACK 60b64682ba338a114d1b27d52fcd6684c44d6be6 not passing tests are clearly not linked to the PR

**Review by Fuzzbawls — APPROVED** [review]
ACK 60b64682ba338a114d1b27d52fcd6684c44d6be6

**Review by Liquid369 — APPROVED** [review]
tACK 60b64682ba338a114d1b27d52fcd6684c44d6be6

**Pull request merged**